Include/internal/pycore_critical_section.h

```diff
@@ -0,0 +1,237 @@
+#ifndef Py_INTERNAL_CRITICAL_SECTION_H
+#define Py_INTERNAL_CRITICAL_SECTION_H
+
+#ifndef Py_BUILD_CORE
+#  error "this header requires Py_BUILD_CORE define"
+#endif
+
+#include "pycore_lock.h"        // PyMutex
+#include "pycore_pystate.h"     // _PyThreadState_GET()
+#include <stdint.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+// Implementation of Python critical sections: helpers to replace the global
+// interpreter lock with per-object locks, while avoiding deadlocks.
+//
```

Comment:
FWIW, your phrasing in https://github.com/python/cpython/pull/111571#discussion_r1384184074 is a bit more clear.  Consider making use of that here, something like:

```suggestion
// Implementation of Python critical sections
//
// Conceptually, critical sections are a deadlock avoidance layer
// on top of per-object locks.  These helpers, in combination with those
// locks, replace our usage of the global interpreter lock to provide
// thread-safety for otherwise thread-unsafe objects, such as dict.
//
```